easybuild/easyconfigs/s/Spark/Spark-2.4.0-Hadoop-2.7-Java-1.8.0_162.eb

```diff
@@ -0,0 +1,29 @@
+easyblock = 'Tarball'
+
+name = 'Spark'
+version = '2.4.0'
+versionsuffix = '-Hadoop-2.7-Java-%(javaver)s'
+
+homepage = 'http://spark.apache.org'
+description = """Spark is Hadoop MapReduce done in memory"""
+
+toolchain = {'name': 'dummy', 'version': ''}
+
+sources = ['%(namelower)s-%(version)s-bin-hadoop2.7.tgz']
+source_urls = [
+    'http://apache.belnet.be/%(namelower)s/%(namelower)s-%(version)s/',
+    'http://www.eu.apache.org/dist/%(namelower)s/%(namelower)s-%(version)s/',
+    'http://www.us.apache.org/dist/%(namelower)s/%(namelower)s-%(version)s/',
+]
+checksums = ['c93c096c8d64062345b26b34c85127a6848cff95a4bb829333a06b83222a5cfa']
+
+dependencies = [('Java', '1.8.0_162')]
```

Comment:
@Falkor I'd strongly prefer if this uses `Java` `1.8` as a dependency, since that makes it compatible with easyconfigs in the `2018b` toolchain generation, see also @migueldiascosta's comment

Note that this implies also renaming the easyconfig file (otherwise tests will fail)

Reply:
OK but now I have to install 1.8.0_181... 
Is there a way to force (or make it 'try' to build) against the installed Java 1.8.0_162 ?

Reply:
@falkor you need to customize the `Java-1.8.eb` file, changing the full version, and install it (with `--rebuild`), which will (re)generate the `.modulerc` file that maps 1.8 to the specific full version

Reply:
@migueldiascosta OK, then sorry for the naive question but when I try to build the custom `Java-1.8.eb` file, it seems to ignore the value of `$MODULEPATH`...

```bash
$> echo $MODULEPATH
$HOME/.local/easybuild/modules/all:/opt/apps/resif/modules:/opt/apps/resif/data/stable/default/modules/all
$> eb --version
This is EasyBuild 3.7.1 (framework: 3.7.1, easyblocks: 3.7.1) on host XXX
$> module show lang/Java/1.8.0_162
-------- /opt/apps/resif/data/stable/default/modules/all/lang/Java/1.8.0_162.lua: --------

$> eb -S Java-1.8.eb
[...]
 * $CFGS1/j/Java/Java-1.8.eb
```
I copied this easyconfigs to reach the following state:
```diff
$> diff -ru $CFGS1/j/Java/Java-1.8.eb Java-1.8.eb
--- $CFGS1/j/Java/Java-1.8.eb   2018-11-22 23:11:07.342444000 +0100
+++ Java-1.8.eb 2018-12-06 08:09:16.650344000 +0100
@@ -9,6 +9,6 @@

 toolchain = {'name': 'dummy', 'version': 'dummy'}

-dependencies = [('Java', '%(version)s.0_181')]
+dependencies = [('Java', '%(version)s.0_162')]

 moduleclass = 'lang'
```
Initial Installation failed despite reporting correct dependencies checks: 
```bash 
$> pwd
$HOME/software/Spark
# Dry run check of dependencies -- seems all good
$> eb ./Java-1.8.eb --robot-paths=$PWD: --rebuild -D
== temporary log file in case of crash /tmp/eb-nFNaPc/easybuild-joubh3.log
Dry run: printing build status of easyconfigs and dependencies
 * [x]  $CFGS1/j/Java/Java-1.8.0_162.eb (module: lang/Java/1.8.0_162)
 * [ ]  $HOME/software/Spark/Java-1.8.eb (module: lang/Java/1.8)
== Temporary log file(s) /tmp/eb-nFNaPc/easybuild-joubh3.log* have been removed.
== Temporary directory /tmp/eb-nFNaPc has been removed.

# Real install fail -- it does not use the appropriate location of the module lang/Java/1.8.0_162
$> == temporary log file in case of crash /tmp/eb-_XWE_j/easybuild-dNkd6k.log
== resolving dependencies ...
== processing EasyBuild easyconfig $HOME/software/Spark/Java-1.8.eb
== building and installing lang/Java/1.8...
== fetching files...
== creating build dir, resetting environment...
== unpacking...
== patching...
== preparing...
== configuring...
== building...
== testing...
== installing...
== taking care of extensions...
== postprocessing...
== sanity checking...
== cleaning up...
== creating module...
== FAILED: Installation ended unsuccessfully (build directory: $HOME/.local/easybuild/build/Java/1.8/dummy-dummy): build failed (first 300 chars): Expected module file $HOME/.local/easybuild/modules/all/lang/Java/1.8.0_162 not found; Lmod 6.x requires that .modulerc and wrapped module file are in same directory!
== Results of the build can be found in the log file(s) /tmp/eb-_XWE_j/easybuild-Java-1.8-20181206.084933.RHlIG.log
ERROR: Build of $HOME/software/Spark/Java-1.8.eb failed (err: 'build failed (first 300 chars): Expected module file $HOME/.local/easybuild/modules/all/lang/Java/1.8.0_162 not found; Lmod 6.x requires that .modulerc and wrapped module file are in same directory!')

real    0m1.647s
user    0m0.955s
sys     0m0.392s
```
I guess from the message it is **really** mandatory to have lang/Java/1.8.0_162 and lang/Java/1.8 at the same prefixed directory... 

Of course, if I enforce the rebuild of the `lang/Java/1.8.0_162` also into my local Easybuild directory, it works -- see the uploaded test report. 
I just wanted to know if that was a desired state (and sorry it it reopens what I imagine could have been long discussions).

Reply:
@Falkor Which EasyBuild version are you using here? We fixed some problems with module wrapper support (which is exactly what `Java/1.8` is) in EasyBuild v3.7.1 .

Reply:
@boegel 3.7.1  (see extract of `eb --version`)

Reply:
@Falkor OK, sorry, I overlooked the actual error message that you shared.

With Lmod 6.x, it's indeed mandatory that both `Java/1.8` and the `Java` module it is wrapping are installed in the same installation prefix.
That's not required with Lmod 7.x, and a sufficiently recent Lmod 7 (i.e. 7.7.38 or newer) also allows using `Lua` syntax for the wrapper.

Long story short: you should look into updating Lmod. ;)